code/modules/shuttle/shuttle_weapons/weapon_controller.dm

```diff
@@ -0,0 +1,319 @@
+		return
+	// Update UI
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		var/user_ref = REF(user)
+		var/is_living = isliving(user)
+		// Ghosts shouldn't count towards concurrent users, which produces
+		// an audible terminal_on click.
+		if(is_living)
+			concurrent_users += user_ref
+		// Turn on the console
+		if(length(concurrent_users) == 1 && is_living)
+			playsound(src, 'sound/machines/terminal_on.ogg', 25, FALSE)
+			use_power(active_power_usage)
+			//Show camera static to the first viewer, since it hides potential mess ups with scaling and viewing dead ships.
+			show_camera_static()
+		// Register map objects
+		user.client.register_map_obj(cam_screen)
+		user.client.register_map_obj(cam_plane_master)
+		user.client.register_map_obj(cam_background)
+		// Open UI
+		ui = new(user, src, "WeaponConsole")
+		ui.open()
+
+/obj/machinery/computer/weapons/ui_data(mob/user)
+	var/list/data = list()
+	var/obj/docking_port/mobile/connected_port = SSshuttle.getShuttle(shuttle_id)
+	data["selectedShip"] = selected_ship_id
+	data["weapons"] = list()
+	data["ships"] = list()
+	if(!connected_port)
+		log_shuttle("Weapons console linked to [shuttle_id] could not locate a connected port using SSshuttle system.")
+		return data
+	var/datum/ship_datum/our_ship = SSbluespace_exploration.tracked_ships[shuttle_id]
+	if(!our_ship || !our_ship.combat_allowed)
+		return data
+	//Enemy Ships
+	for(var/ship_id in SSbluespace_exploration.tracked_ships)
+		var/datum/ship_datum/ship = SSbluespace_exploration.tracked_ships[ship_id]
+		//Shooting ourself, 🤔
+		if(ship.mobile_port_id == shuttle_id)
+			continue
+		//Ignore ships that are on different z-levels#
+		var/obj/target_port = SSshuttle.getShuttle(ship_id)
+		if(!target_port || target_port.z != connected_port.z)
+			continue
+		if(!ship.combat_allowed)
+			continue
+		var/datum/faction/their_faction = ship.ship_faction
+		var/list/other_ship = list(
+			id = ship_id,
+			name = ship.ship_name,
+			faction = ship.ship_faction,
+			health = ship.integrity_remaining,
+			maxHealth = ship.max_ship_integrity * SHIP_INTEGRITY_FACTOR,
+			critical = ship.critical,
+			//If they consider us hostile (AI works on if we consider them hostile policy, but that is confusing for players to be shot at by 'friendly' ships).
+			hostile = check_faction_alignment(ship.ship_faction, our_ship.ship_faction) == FACTION_STATUS_HOSTILE || (their_faction.type in our_ship.rogue_factions),
+		)
+		data["ships"] += list(other_ship)
+	var/list/turfs = connected_port.return_turfs()
+	//Weapons
+	for(var/turf/T as() in turfs)
+		for(var/obj/machinery/shuttle_weapon/weapon in T)
+			var/list/active_weapon = list(
+				id = weapon.unique_id,
+				name = weapon.name,
+				cooldownLeft = max(weapon.next_shot_world_time - world.time, 0),
+				cooldown = weapon.cooldown,
+				inaccuracy = weapon.innaccuracy,
+			)
+			data["weapons"] += list(active_weapon)
```

Comment:
Why do you do this when you have the `weapon_weakrefs` (not yet implemented, sadly) list?